ruby/ql/docs/flow_summaries.md

```diff
@@ -0,0 +1,205 @@
+arbitrary objects as keys). Elements can also have an unknown index, which means
+we know the element exists in the collection but we don't know where.
+
+Many of the specifiers have an optional suffix `!`. If this suffix is used then
+the specifier excludes elements at unknown indices. Otherwise, these are
+included by default.
+
+### Specifiers
+
+#### `?`
+An element at an unknown index.
+
+#### `any`
+An element at any known or unknown index.
+
+#### `<integer>`, `<integer>!`
+An element at the index given by the integer.
+
+#### `<integer>..`, `<integer>..!`
+Any element at a known index greater or equal to the integer.
+
+#### `<string>`, `<string>!`
+An element at the index given by string. The string should match the result of
+`serialize()` on the `ConstantValue` that represents the index. This is
+typically something like `foo` for the string key `"foo"` and `:foo` for the
+symbol `:foo`.
+
+## `Field`
+A "field" in the object. In practice this refers to a value stored in an
+instance variable in the object. The only valid specifier is `@<string>`, where
+`<string>` is the name of the instance variable. Currently we assume that a
+setter call such as `x.foo = bar` means there is a field `foo` in `x`, backed by
+an instance variable `@foo`.
+
+For example, the access path `Argument[0].Field[@foo]` would refer to the value
+`"foo"` in
+
+```rb
+x = SomeClass.new
+x.foo = "foo"
+some_call(x)
+```
+
+## `WithElement`
+This component restricts the set of elements that are included in the preceding
+access path to to those at a specific set of indices. The specifiers are the
+same as those for `Element`.
+
+When used in an input path this component has the effect of copying
+all relevant elements from the input to the output. For example, in the
+following summary:
+
+```ql
+input = "Argument[0].WithElement[1, 2]" and
+output = "ReturnValue"
+```
+
+any data in indices 1 and 2 of the first argument will be copied to indices 1
+and 2 of the return value. We use this in many Hash summaries that return the
+receiver, in order to preserve any data stored in it. For example, the summary
+for `Hash#to_h` is
+
+```ql
+input = "Argument[self].WithElement[any]" and
+output = "ReturnValue" and
+preservesValue = true
+```
+
+TODO: I've not seen this component used in an output path; I don't know if it makes
+sense to do so, or what meaning it would have.
+
+## `WithoutElement`
+This component is used to exclude certain elements from the set included in the
+preceding access path. It takes the same specifiers as `WithElement` and
+`Element`.
+
+When used in an input path this component has the effect of excluding the
+relevant elements when copying from input to output. For example in the
+following summary:
+
+```ql
+input = "Argument[0].WithoutElement[0]" and
+output = "ReturnValue"
+```
+
+any data in any index of the first argument will be copied to the return value,
+with the exception of data at index 0.
```

Comment:
A prominent use case of `WithoutElement` is when you want to model a method that removes (a set of) element(s) from a collection. For example, if we want to model a method that removes the first element of the receiver collection, we can do so via

```ql
input = "Argument[self].WithoutElement[0]" and
output = "Argument[self]"
```

where both the input and the output refer to the same argument. The data-flow library will then prevent use-use flow from any corresponding receivers:

```rb
a[0] = taint
a[1] = taint
a.remove_first # use-use flow from `a` to the `a` below will be blocked, but there will still be flow from `[post-update] a` to `a` below
sink(a[0]) # good
sink(a[1]) # bad
```

Reply:
A good example of this is something like `Hash#except`. If we see a call like `hash.except(:foo)`, we would like that to map to a flow summary with
```
input = Argument[self].WithoutElement[":foo"].WithElement[any]
output = ReturnValue
```
I.e. data can flow out of the call, but only when it is stored in _some_ element that is not known to be at index `:foo`.

Reply:
This would also incorrectly include flow for `sink(source("d").s53())`.

Reply:
Yes, that looks good. `WithoutElement[x]` just says that data must _not_ be stored inside a collection at index `x`, not that data _must_ be stored inside a collection.

This is also the reason why something like a `reverse` method needs a summary like
```ql
input = "Argument[self].WithElement[any]" and
output = "ReturnValue"
```
and not
```ql
input = "Argument[self]" and
output = "ReturnValue"
```